worker/src/polkadex_db.rs

```diff
@@ -0,0 +1,145 @@
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::thread;
+use std::thread::JoinHandle;
+
+use codec::Encode;
+use log::error;
+use rocksdb::{DB, DBWithThreadMode, Error as RocksDBError, Error, IteratorMode, Options, SingleThreaded};
+
+use polkadex_primitives::types::{Order, SignedOrder};
+
+///
+/// Polkadex Orderbook Mirror Documentation
+/// The backend DB is RocksDb
+/// Orders are stored as (counter,SignedOrder)
+/// where SignedOrder contains Order, counter and signature of enclave
+///
+///
+///
+///
+
+use crate::constants::ORDERBOOK_DB_FILE;
+
+static ORDERBOOK_MIRROR: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+
+pub struct RocksDB {
+    db: DBWithThreadMode<SingleThreaded>,
+}
+
+pub enum PolkadexDBError {
+    UnableToLoadPointer,
+    UnableToRetrieveValue,
+    ErrorWritingToDB,
+    UnableToDeseralizeValue,
+    ErrorDeleteingKey,
+}
+
+pub trait KVStore {
+    /// Loads the DB from file on disk
+    fn initialize_db(create_if_missing_db: bool) -> Result<(), RocksDBError>;
+    fn load_orderbook_mirror() -> Result<&'static Mutex<RocksDB>, PolkadexDBError>;
+    fn write(order_uid: Vec<u8>, signed_order: SignedOrder) -> JoinHandle<Result<(), PolkadexDBError>>;
+    fn find(k: Vec<u8>) -> Result<Option<SignedOrder>, PolkadexDBError>;
+    fn delete(k: Vec<u8>) -> JoinHandle<Result<(), PolkadexDBError>>;
+    fn read_all() -> Result<Vec<SignedOrder>,PolkadexDBError>;
+}
+
+impl KVStore for RocksDB {
+    fn initialize_db(create_if_missing_db: bool) -> Result<(), RocksDBError> {
+        let mut opts = Options::default();
+        opts.create_if_missing(create_if_missing_db);
+
+        let db = DB::open(&opts, ORDERBOOK_DB_FILE)?;
+        let storage_ptr = Arc::new(Mutex::<RocksDB>::new(RocksDB { db }));
+        let ptr = Arc::into_raw(storage_ptr);
+        // FIXME: Do we really need SeqCst here?, RocksDB already takes care of concurrent writes.
+        ORDERBOOK_MIRROR.store(ptr as *mut (), Ordering::SeqCst);
+        Ok(())
+    }
+
+    fn load_orderbook_mirror() -> Result<&'static Mutex<RocksDB>, PolkadexDBError> {
+        let ptr = ORDERBOOK_MIRROR.load(Ordering::SeqCst) as *mut Mutex<RocksDB>;
+        if ptr.is_null() {
+            println!(" Unable to load the pointer");
+            return Err(PolkadexDBError::UnableToLoadPointer);
+        } else {
+            Ok(unsafe { &*ptr })
+        }
+    }
+    fn write(order_uid: Vec<u8>, signed_order: SignedOrder) -> JoinHandle<Result<(), PolkadexDBError>> {
+        thread::spawn(move || -> Result<(), PolkadexDBError> {
```

Comment:
Maybe it would make more sense to start a new thread and load the orderbook mirror outside of this function? 
Doing it this way, you're implying that for every write the pointer needs to be loaded. In the future, there might be cases where read and writes  would want to be done.

It would also be better to understand (just seeing a RocksDB::write() does not imply a threading & pointer loading).

Reply:
Resolved.

Reply:
I have moved the threading and pointer outside for both write and delete